File: apps/board/backend/app/models/lead.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class LeadStatus(str, Enum):
    nuovo = "nuovo"
    contatto = "contatto"
    proposta = "proposta"
    chiuso_vinto = "chiuso_vinto"
    chiuso_perso = "chiuso_perso"


class LeadSource(str, Enum):
    k_bot = "k_bot"
    contact_form = "contact_form"
    newsletter = "newsletter"
    referral = "referral"
    outbound = "outbound"
    other = "other"


class LeadBase(BaseModel):
    contact_id: Optional[UUID] = None
    title: str
    description: Optional[str] = None
    status: LeadStatus = LeadStatus.nuovo
    source: LeadSource = LeadSource.other
    value_eur: Optional[Decimal] = None
    probability: int = Field(default=50, ge=0, le=100)
    next_action_at: Optional[datetime] = None
    next_action_note: Optional[str] = None
    kbot_session_id: Optional[UUID] = None


class LeadCreate(LeadBase):
    pass


class LeadUpdate(BaseModel):
    contact_id: Optional[UUID] = None
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[LeadStatus] = None
    source: Optional[LeadSource] = None
    value_eur: Optional[Decimal] = None
    probability: Optional[int] = Field(default=None, ge=0, le=100)
    next_action_at: Optional[datetime] = None
    next_action_note: Optional[str] = None
    kbot_session_id: Optional[UUID] = None
    closed_at: Optional[datetime] = None


class LeadRead(LeadBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    created_at: datetime
    updated_at: datetime
    closed_at: Optional[datetime] = None

File: apps/board/web/src/types/lead.ts
export type LeadStatus =
  | "nuovo"
  | "contatto"
  | "proposta"
  | "chiuso_vinto"
  | "chiuso_perso";

export type LeadSource =
  | "k_bot"
  | "contact_form"
  | "newsletter"
  | "referral"
  | "outbound"
  | "other";

export interface Lead {
  id: string;
  contact_id: string | null;
  title: string;
  description: string | null;
  status: LeadStatus;
  source: LeadSource;
  value_eur: number | null;
  probability: number;
  next_action_at: string | null;
  next_action_note: string | null;
  kbot_session_id: string | null;
  created_at: string;
  updated_at: string;
  closed_at: string | null;
}

export interface LeadCreatePayload {
  contact_id?: string | null;
  title: string;
  description?: string | null;
  status?: LeadStatus;
  source?: LeadSource;
  value_eur?: number | null;
  probability?: number;
  next_action_at?: string | null;
  next_action_note?: string | null;
}

export type LeadUpdatePayload = Partial<LeadCreatePayload> & {
  closed_at?: string | null;
};

export interface Contact {
  id: string;
  company: string | null;
  person_name: string | null;
  email: string | null;
  phone: string | null;
  notes: string | null;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface ContactCreatePayload {
  company?: string | null;
  person_name?: string | null;
  email?: string | null;
  phone?: string | null;
  notes?: string | null;
  tags?: string[];
}
